### DISCUSSION TITLE
How to set span name in AspNetCoreInstrumentation and HttpClientInstrumentation

### DISCUSSION THREAD

**Discussion by ajitbhalekar** [open-telemetry/opentelemetry-dotnet#4292] [Q&A] [answered]
I want to set span in specific format when there is specific header present in request. How can I set that.  This is some sample I am trying out.

 _ = services
            .AddOpenTelemetry()
            .WithTracing(builder => builder
            .SetSampler(new AlwaysOnSampler())
                .AddAspNetCoreInstrumentation(options =>
                 {
                     const string incomingSpanName = "IncomingSpanName";
                     options.EnrichWithHttpRequest = (activity, httpRequest) =>
                     {
                        if (httpRequest.Headers.TryGetValue("x-ms-correlation-request-id", out StringValues clientActionHeaders))
                         {
                             string displayName = "incomingSpanName";
                             activity.AddTag(incomingSpanName, displayName);

                             // Ideally this should be sufficient - however, OT ends up resetting this value to the action's route template
                             activity.DisplayName = displayName;
                         }
                     };

                     options.EnrichWithHttpResponse = (activity, response) =>
                     {
                         // Hack to get the value added to the tag to setup the display name
                         object? incomingSpan = activity.GetTagItem(incomingSpanName);
                         if (incomingSpan != null)
                         {
                             activity.DisplayName = "incomingSpan";
                         }
                     };
                 })
                .AddHttpClientInstrumentation(options =>
                  options.EnrichWithHttpRequestMessage = (activity, httpRequest) =>
                  {
                      if (httpRequest.Headers.TryGetValues("x-ms-correlation-request-id", out IEnumerable<string>? clientActionHeaders))
                      {
                          string clientActionName = string.Join(".", clientActionHeaders);
                          if (!string.IsNullOrWhiteSpace(clientActionName))
                          {
                              activity.SetTag("CallerClientRequestId", clientActionName);
                          }
                      }
                  })
                .AddSource(ApiQoSActivitySource.DefaultName)
                .AddConsoleExporter())

**Comment by vishweshbankwar** [comment]
This is an open issue - https://github.com/open-telemetry/opentelemetry-dotnet/issues/3977

**Reply by vishweshbankwar** [reply]
@ajitbhalekar try the workaround I shared in #3977

**Comment by ajitbhalekar** [comment]
Thanks @vishweshbankwar for reply.
To test it I have tried something like below

` _ = services
            .AddOpenTelemetry()
            .WithTracing(builder => builder
                .SetSampler(new AlwaysOnSampler())
                .AddAspNetCoreInstrumentation(o =>
                {
                    o.EnrichWithHttpResponse = (activity, response) =>
                    {
                        string displayName = $"Test.Incoming";

                        activity.DisplayName = displayName;
                    };
                })
                .AddConsoleExporter()`

But the output was not as per expected.
`Activity.TraceId:            c1e400151864ae36dce6c0de0aedf71f
Activity.SpanId:             696d5e1cb9a55f88
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Microsoft.AspNetCore
Activity.DisplayName:        /swagger/v1/swagger.json
Activity.Kind:               Server
Activity.StartTime:          2023-03-17T18:48:24.6738489Z
Activity.Duration:           00:00:00.1255971
Activity.Tags:
    net.host.name: localhost
    net.host.port: 51942
    http.method: GET
    http.scheme: https
    http.target: /swagger/v1/swagger.json
    http.url: https://localhost:51942/swagger/v1/swagger.json
    http.flavor: 2.0
    http.user_agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36 Edg/111.0.1661.43
    http.status_code: 200
Resource associated with Activity:
    service.name: unknown_service:UsageIngestion`

**Reply by vishweshbankwar** [reply]
@ajitbhalekar - Could you share a minimal repro app? I just tried this and here is the output

![image](https://user-images.githubusercontent.com/17655898/226000802-7727361b-57c5-44fe-99c1-8fa79f6b570b.png)

My set up
```
builder.Services.AddOpenTelemetry().WithTracing(builder =>
{
    builder.AddAspNetCoreInstrumentation(o => o.EnrichWithHttpResponse = (activity, _) =>
    {
        activity.DisplayName = "MyDisplayName";
    })
    .AddConsoleExporter();
});

**Reply by ajitbhalekar** [reply]
I have tried with minimal changes above.  but the result is not as per expected.

This are the packages I have added in my our .Net7 application.

` <PackageReference Update="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.11" />
    <PackageReference Update="OpenTelemetry.Extensions.Hosting" Version="1.4.0-rc.2" />
    <PackageReference Update="OpenTelemetry.Exporter.Console" Version="1.4.0-rc.3" />
    <PackageReference Update="OpenTelemetry.Extensions.DependencyInjection" Version="1.4.0-rc.2" />
    <PackageReference Update="OpenTelemetry.Instrumentation.Runtime" Version="1.1.0-beta.4" />
    <PackageReference Update="OpenTelemetry.Instrumentation.Process" Version="1.0.0-alpha.5" />
    <PackageReference Update="OpenTelemetry.Exporter.Geneva" Version="1.3.1" />`

**Comment by ajitbhalekar** [comment, marked as answer]
Finally found the root cause and thanks @vishweshbankwar for having a close look at my package reference and driving to the solution. 

Actually the issue over here the package reference added for 
OpenTelemetry.Extensions.Hosting
OpenTelemetry.Exporter.Console
OpenTelemetry.Instrumentation.Runtime
OpenTelemetry.Instrumentation.Process.

So after updating my packages with stable version and correcting the available versions for Runtime and Process, it has worked as   per expected. 

My updated package references:
`    <PackageReference Update="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.14" />
    <PackageReference Update="OpenTelemetry.Extensions.Hosting" Version="1.4.0" />
    <PackageReference Update="OpenTelemetry.Exporter.Console" Version="1.4.0" />
    <PackageReference Update="OpenTelemetry.Instrumentation.Runtime" Version="1.1.0-rc.2" />
    <PackageReference Update="OpenTelemetry.Instrumentation.Process" Version="0.5.0-beta.2" />
    <PackageReference Update="OpenTelemetry.Exporter.Geneva" Version="1.4.0" />`
